consensus/consortium/v2/consortium.go

```diff
@@ -837,11 +848,23 @@ func (c *Consortium) Prepare(chain consensus.ChainHeaderReader, header *types.He
 		// at the start of every epoch.
 		if c.IsTrippEffective(chain, header) {
 			if c.IsPeriodBlock(chain, header) {
-				extraData.CheckpointValidators = checkpointValidator
+				extraData.CheckpointValidators = checkpointValidators
+			} else {
+				// checkpointValidators get from contract is nil at epoch block; hence,
+				// checkpointValidators is reassigned by the candidate list from snapshot,
+				// which is used in encodingValidatorBitSet step.
+				checkpointValidators = snap.ValidatorsWithBlsPub
+			}
```

Comment:
I've just realized this is incorrect, we must always use snap.ValidatorsWithBlsPub. At period block, we don't switch to new block producer set right way but wait until len(blockProducer) / 2 (logic in snapshot). So the block producer here must be the indices of block producers in the old set in snap.ValidatorsWithBlsPub not the new set in extradata.

Reply:
This is an incorrect comment